### PR TITLE
New Gallery

### PR DESCRIPTION
**Pull request by jisqyv**
Note: This PR also includes the contents of #2285

### CODE REVIEW

**Comment by jisqyv** [comment]
This PR includes the removal of the old gallery. Ultimately, before merging I will reduce it to 2 commits. One that removes the old gallery and another that adds the new gallery.

**Comment by jisqyv** [comment]
@ewpatton Finally ready to review!

**Comment by jisqyv** [comment]
@ewpatton Note: Should be rebased as two commits.

File: appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidProjectService.java
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidProjectService.java:1085]: A minor improvement would be to use a TreeSet and request an iterator starting at the point where the project name collides and walk the (sorted) project names from there. Granted, most people will only have a few projects max (average is ~5 per user).

File: appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidProjectService.java
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidProjectService.java:1062]: We should make a note for the future that this function should only be called for servers we trust as using the Content-Length header could be problematic with untrustworthy servers.

**Review by ewpatton — CHANGES_REQUESTED** [review]

File: appinventor/appengine/src/com/google/appinventor/client/AdminUserList.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/client/AdminUserList.java]: Done

File: appinventor/appengine/src/com/google/appinventor/server/ProjectServiceImpl.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/server/ProjectServiceImpl.java]: Done. But I put an expanded comment in ObjectifyStorageIo where the project type is looked up (the code there is commented out and a constant returned).

File: appinventor/appengine/src/com/google/appinventor/server/ProjectServiceImpl.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/server/ProjectServiceImpl.java]: Done. I changed most references to "New Gallery" to just Gallery. Exception: I didn't change the newGalleryId and related flags in Ode. But may at some future date.

File: appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/GalleryToken.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/GalleryToken.java:16]: I do not want to change this. The token gets encrypted, which means that if it is one byte over the blocksize of the cipher, a whole new block is used. Protocol buffers use an efficient representation for small integers that I am taking advantage of here. Finally, this timestamp is ephemeral. It is never saved to permanent storage where it may be confused. This hack is limited to this code and the corresponding "verifytoken" routine on the Gallery side.

File: appinventor/appengine/src/com/google/appinventor/server/CrashReport.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/server/CrashReport.java:63]: I added a Javadoc. It is not used here but is used in the Stand Alone and Scale-able code (where I am using log4j). This code was carried over as an artifact of how I removed the gallery, to be consistent with how I removed it a while ago from the stand alone and scale-able code. Leaving it here makes the merge of this code into the standalone and scale-able code easier.

File: appinventor/appengine/src/com/google/appinventor/client/DesignToolbar.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/client/DesignToolbar.java]: Yes, fixed. Good catch!

File: appinventor/appengine/src/com/google/appinventor/client/DesignToolbar.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/client/DesignToolbar.java:394]: Agreed. Left alone.

File: appinventor/appengine/src/com/google/appinventor/server/ProjectServiceImpl.java
Comment (ewpatton) [reply appinventor/appengine/src/com/google/appinventor/server/ProjectServiceImpl.java]: I agree that it's acceptable to keep the old/new dichotomy in this case.

File: appinventor/appengine/src/com/google/appinventor/client/TopPanel.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/client/TopPanel.java]: I didn't touch this code in this PR, this is old cruft in the master branch, but I did remove it.

File: appinventor/appengine/src/com/google/appinventor/client/Ode.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/client/Ode.java]: I refactored this code a bit. It doesn't make sense to have the newGalleryId and templateId at the same time. But if someone constructs such a thing, the template takes precidence.

File: appinventor/appengine/src/com/google/appinventor/client/Ode.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/client/Ode.java:690]: Done.

File: appinventor/appengine/src/com/google/appinventor/client/Ode.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/client/Ode.java:2196]: Same comment as above, this is an artifact from the standalone/scale-able code (which has support for licenses and expiration dates).

File: appinventor/appengine/src/com/google/appinventor/server/storage/ObjectifyStorageIo.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/server/storage/ObjectifyStorageIo.java]: This was what was present prior to the old gallery being integrated, so that is what I did. However I changed it now.

Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/server/storage/ObjectifyStorageIo.java]: ditto

File: appinventor/appengine/src/com/google/appinventor/server/storage/ObjectifyStorageIo.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/server/storage/ObjectifyStorageIo.java]: ditto

File: appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidProjectService.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidProjectService.java]: Done (under a DEBUG if statement).

File: appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidProjectService.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidProjectService.java]: ditto

File: appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidProjectService.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidProjectService.java]: Done. I didn't change the code, I just swapped out the ArrayList for a TreeSet. The TreeSet (well, really TreeMap) implementation of "contains" does binary search, which is more efficient. There is overhead when the TreeSet is created, but that is done once while the "contains" may be called multiple times if we have multiple collisions. Note: New code tested, complete with forced collisions.

File: appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidProjectService.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidProjectService.java:1068]: Understood, but not changed for now.

**Comment by jisqyv** [comment]
@ewpatton If you approve this change, please approve the review and mark it LGTM, but do not merge. I'm going to want to merge it manually to preserve the signatures on the commits.

File: appinventor/appengine/src/com/google/appinventor/client/TopPanel.java
Comment (ewpatton) [reply appinventor/appengine/src/com/google/appinventor/client/TopPanel.java]: Right. Since we aren't using it currently there's no reason to keep it around.

File: appinventor/appengine/src/com/google/appinventor/client/Ode.java
Comment (ewpatton) [reply appinventor/appengine/src/com/google/appinventor/client/Ode.java:2196]: Can we just put it in the rados branch then? (or whatever the appropriate branch is). Given this affects nothing in the open source I'd rather not pull it into the open source.

File: appinventor/appengine/src/com/google/appinventor/client/Ode.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/client/Ode.java]: Oh. I got your point. But when I looked at the code it just didn't make sense as written, so I rewrote it a bit. The new version doesn't reference newGalleryId before it is set.

File: appinventor/appengine/src/com/google/appinventor/server/CrashReport.java
Comment (ewpatton) [reply appinventor/appengine/src/com/google/appinventor/server/CrashReport.java:63]: It might be better if we can either use the Java Logging framework or the Apache framework. I don't really see the point of mixing the two frameworks.

File: appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/GalleryToken.java
Comment (ewpatton) [reply appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/GalleryToken.java:16]: Fair enough.

File: appinventor/appengine/src/com/google/appinventor/client/Ode.java
Comment (ewpatton) [reply appinventor/appengine/src/com/google/appinventor/client/Ode.java]: Ok. Maybe it's an issue on my end as GitHub seems to be showing the older version.

**Comment by jisqyv** [comment]
@ewpatton I've remoted the code that isn't used here.

**Comment by wxbit** [comment]
Is there a demo? Or images.

**Comment by jisqyv** [comment]
@wxbit Yes, login to http://ai2-test.appinventor.mitedu

File: appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectToolbar.java
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectToolbar.java:204]: Maybe a more appropriate message in OdeMessages?

File: appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectToolbar.java
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectToolbar.java:229]: OdeMessages

File: appinventor/appengine/src/com/google/appinventor/client/OdeMessages.java
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/OdeMessages.java:4781]: I thought this used to be in the branding branch. Just want to check that moving it here is intentional.

**Review by ewpatton — CHANGES_REQUESTED** [review]
A few minor things to address.

File: appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectToolbar.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectToolbar.java]: Oops. Fixed.

File: appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectToolbar.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectToolbar.java]: Fixed.

File: appinventor/appengine/src/com/google/appinventor/client/OdeMessages.java
Comment (jisqyv) [reply appinventor/appengine/src/com/google/appinventor/client/OdeMessages.java]: Nope, not intentional and removed.

**Comment by jisqyv** [comment]
@ewpatton Last changes made.

**Review by ewpatton — APPROVED** [review]
LGTM

**Pull request merged**
